src/components/tl/mlx5/tl_mlx5_dm.c

```diff
@@ -0,0 +1,180 @@
+    }
+    return UCC_OK;
+}
+
+static void ucc_tl_mlx5_dm_chunk_init(ucc_mpool_t *mp,        //NOLINT
+                                      void *obj, void *chunk) //NOLINT
+{
+    ucc_tl_mlx5_dm_chunk_t *c = (ucc_tl_mlx5_dm_chunk_t *)obj;
+    ucc_tl_mlx5_team_t *    team =
+        ucc_container_of(mp, ucc_tl_mlx5_team_t, dm_pool);
+    const size_t memic_chunk = UCC_TL_MLX5_DM_CHUNK_SIZE;
+    c->offset                = (ptrdiff_t)team->oob_req;
+    team->oob_req            = PTR_OFFSET(team->oob_req, memic_chunk);
+}
+
+static void ucc_tl_mlx5_dm_chunk_release(ucc_mpool_t *mp, void *chunk) //NOLINT
+{
+    ucc_free(chunk);
+}
+
+ucc_mpool_ops_t ucc_tl_mlx5_dm_ops = {.chunk_alloc = ucc_tl_mlx5_dm_chunk_alloc,
+                                      .chunk_release =
+                                          ucc_tl_mlx5_dm_chunk_release,
+                                      .obj_init    = ucc_tl_mlx5_dm_chunk_init,
+                                      .obj_cleanup = NULL};
+
+void ucc_tl_mlx5_dm_cleanup(ucc_tl_mlx5_team_t *team)
+{
+    if (!team->dm_ptr) {
+        return;
+    }
+
+    ucc_mpool_cleanup(&team->dm_pool, 1);
+
+    ibv_dereg_mr(team->dm_mr);
+    if (UCC_TL_MLX5_TEAM_LIB(team)->cfg.dm_host) {
+        ucc_free(team->dm_ptr);
+    } else {
+        ibv_free_dm(team->dm_ptr);
+    }
+}
+
+ucc_status_t ucc_tl_mlx5_dm_alloc_reg(struct ibv_context *ib_ctx,
+                                      struct ibv_pd *pd, int dm_host,
+                                      size_t buf_size, size_t *buf_num_p,
+                                      struct ibv_dm **ptr, struct ibv_mr **mr,
+                                      ucc_base_lib_t *lib)
+{
+    struct ibv_dm *           dm_ptr = NULL;
+    struct ibv_mr *           dm_mr;
+    struct ibv_device_attr_ex attr;
+    struct ibv_alloc_dm_attr  dm_attr;
+    int                       max_chunks_to_alloc, min_chunks_to_alloc, i;
+
+    if (dm_host) {
+        max_chunks_to_alloc = (*buf_num_p == UCC_ULUNITS_AUTO) ? 8 : *buf_num_p;
+        dm_attr.length      = max_chunks_to_alloc * buf_size;
+        dm_ptr              = ucc_malloc(dm_attr.length, "memic_host");
+        if (!dm_ptr) {
+            tl_error(lib, " memic_host allocation failed");
+            return UCC_ERR_NO_MEMORY;
+        }
+
+        dm_mr = ibv_reg_mr(pd, dm_ptr, dm_attr.length,
+                           IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_WRITE);
+        if (!dm_mr) {
+            tl_error(lib, "failed to reg host memory");
+            ucc_free(dm_ptr);
+            return UCC_ERR_NO_MESSAGE;
+        }
+        *buf_num_p = max_chunks_to_alloc;
+    } else {
+        attr.comp_mask = 0;
+        if (ibv_query_device_ex(ib_ctx, NULL, &attr)) {
+            tl_error(lib, "failed to query device (errno=%d)", errno);
+            return UCC_ERR_NO_MESSAGE;
+        }
+        if (!attr.max_dm_size) {
+            tl_error(lib, "device doesn't support dm allocation");
```

Comment:
For now, I think this is preferable to have an error and not silently fallback to device memory
The option to use host memory in place of device memory is mainly for debug reason, since in this case, the performance really drops
If you don't think this way lmk and I'll make the change